phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java

```diff
@@ -160,17 +241,312 @@ public void close() throws IOException {
+                        int tableId = ttlCache.addTable(m);
+                        // map the row-prefix to the tableId using prefix index.
+                        index.put(m.getPrefix(), tableId);
+                        if (LOGGER.isDebugEnabled()) {
+                            LOGGER.debug(String.format("Updated index : " + m.toString()));
+                        }
+
+                    }
+                });
+            } catch (Exception e) {
+                LOGGER.error(String.format("Failed to read from catalog: " + e.getMessage()));
+            } finally {
+                LOGGER.info(String.format("PartitionedTableTTLTracker " +
+                                "index-entries = %d, cache-entries = %d for region = %s",
+                        index.getValidPrefixes(), ttlCache.getNumTablesInCache(),
+                        CompactionScanner.this.store.getRegionInfo().getEncodedName()));
+            }
+
+        }
+        @Override
+        public void setTTL(Cell firstCell) {
+
+            boolean matched = false;
+            TableTTLInfo tableTTLInfo = null;
+            List<Integer> pkPositions = null;
+            long rowTTLInSecs = ttl;
+            long searchOffset = -1;
+            int pkPosition = startingPKPosition;
+            try {
+                if (ttlCache.getNumTablesInCache() == 0) {
+                    // case: no views in the hierarchy had TTL set
+                    // Use the default TTL
+                    this.rowContext = new RowContext();
+                    this.rowContext.setTTL(rowTTLInSecs);
+                    return;
+                }
+                // pkPositions holds the byte offsets for the PKs of the base table
+                // for the current row
+                pkPositions = isIndexTable ?
+                        (isSalted ?
+                                Arrays.asList(0, 1) :
+                                Arrays.asList(0)) :
+                        rowKeyParser.parsePKPositions(firstCell);
+                // The startingPKPosition was initialized(constructor) in the following manner
+                // case multi-tenant + salted + index-table  => startingPKPosition = 1
+                // case multi-tenant + salted => startingPKPosition = 2, 1
+                // case multi-tenant + index-table => startingPKPosition = 0
+                // case salted + index-table => startingPKPosition = 1
+                // case salted => startingPKPosition = 1
+                // case index-table => startingPKPosition = 0
+                // case multi-tenant => startingPKPosition = 1, 0
+                int offset = pkPositions.get(pkPosition);
+                byte[] rowKey = CellUtil.cloneRow(firstCell);
+                // Search using the starting offset (startingPKPosition offset)
+                Integer tableId = index.getTableIdWithPrefix(rowKey, offset);
+                if ((tableId == null) && (!isIndexTable) && (isMultiTenant || isSalted)) {
```

Comment:
@jpisaac What if my partition key is a prefix of tenantid ? I define a ttl on global view and a ttl on a tenant view created on a different global view. Basically there is one ttl on "KP1" and another on "KP1Tenant#KP2"